plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -910,10 +913,16 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
         IcebergOptimizeHandle optimizeHandle = (IcebergOptimizeHandle) executeHandle.getProcedureHandle();
         Table icebergTable = transaction.table();
 
-        // paths to be deleted
-        Set<DataFile> scannedFiles = splitSourceInfo.stream()
-                .map(DataFile.class::cast)
-                .collect(toImmutableSet());
+        // files to be deleted
+        ImmutableSet.Builder<DataFile> scannedDataFilesBuilder = ImmutableSet.builder();
+        ImmutableSet.Builder<DeleteFile> scannedDeleteFilesBuilder = ImmutableSet.builder();
+        splitSourceInfo.stream().map(DataFileWithDeleteFiles.class::cast).forEach(dataFileWithDeleteFiles -> {
+            scannedDataFilesBuilder.add(dataFileWithDeleteFiles.getDataFile());
+            scannedDeleteFilesBuilder.addAll(filterDeleteFiles(dataFileWithDeleteFiles.getDeleteFiles(), optimizeHandle.isTupleDomainAll()));
```

Comment:
If you run `filterDeleteFiles` on the set after the loop you can get away without scanning duplicates multiple times.

Reply:
yes but i will have to iterate through those files again. Do you think the number of duplicates is that big it makes sense to do it ? I honestly don't know

Reply:
That's fair. I'm fine with either